### PR TITLE
Add getVehicleWheelFrictionState function

### PR DESCRIPTION
**Pull request by onikyannn**
This PR adds `getVehicleWheelFrictionState` function.
Very useful for modifying vehicle sounds and physics.

Usage:
`int getVehicleWheelFrictionState ( vehicle theVehicle, int wheel )`
OOP:
`vehicle:getWheelFrictionState ( int wheel )`

The argument is a wheel number (0/1/2/3).

Function returns number from 0 to 3, where:
0 - normal friction
1 - slip with acceleration (only for driving wheels)
2 - slip without acceleration
3 - locked wheel (on brake on handbrake)

[test_resource](https://github.com/multitheftauto/mtasa-blue/files/5550223/test_resource.zip)

[Demonstration video](https://www.youtube.com/watch?v=qjALFqNnVNk&feature=youtu.be)

**EDIT (04/01/2021):**
[Wiki page](https://wiki.multitheftauto.com/wiki/GetVehicleWheelFrictionState)

### CODE REVIEW

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp [hunk 1]
```diff
@@ -954,6 +956,30 @@ int CLuaVehicleDefs::IsVehicleWheelCollided(lua_State* luaVM)
...
+            lua_pushboolean(luaVM, false);
+    }
+    else
+    {
+        m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());
+        lua_pushboolean(luaVM, false);
+    }
+    return 1;
+}
+
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: Use the new parser instead.
Search for `ArgumentParser` in the code.
IIRC theres a wiki page about it as well (here on github)
Your function args will need to be:  `CVehicle* pVehicle, std::variant<eWheelPosition, unsigned char>`.

**Review by Pirulax — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp (hunk 1 above)
Comment (onikyannn) [reply Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: Thank you for review.
Сhanges applied.
Also moved all checkers to `CLuaVehicleDefs.cpp`.

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp [hunk 2]
```diff
@@ -4109,3 +4111,14 @@ bool CLuaVehicleDefs::SetVehicleModelWheelSize(const unsigned short usModel, con
...
+int CLuaVehicleDefs::GetVehicleWheelFrictionState(CClientVehicle* pVehicle, unsigned char wheel)
+{
+    if (wheel < 0 || wheel > 3)
+        throw std::invalid_argument("Invalid wheel number");
+
+    if (CClientVehicleManager::GetVehicleType(pVehicle->GetModel()) != CLIENTVEHICLE_CAR)
+        throw std::invalid_argument("Invalid vehicle type");
+
+    return pVehicle->GetWheelFrictionState(wheel);
+}
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: Very minor nit, this is not using the editorconfig format. there should be a new line at the end of this file:

```suggestion
}

```

File: Client/sdk/game/CVehicle.h [hunk 3]
```diff
@@ -251,6 +251,7 @@ class CVehicle : public virtual CPhysical
     virtual void SetBikeWheelStatus(BYTE bWheel, BYTE bStatus) = 0;
 
     virtual bool IsWheelCollided(BYTE bWheel) = 0;
+    virtual int GetWheelFrictionState(BYTE bWheel) = 0;
```
Comment (qaisjp) [inline Client/sdk/game/CVehicle.h]: Nit: this is not using the editorconfig format. try this:

```suggestion
    virtual bool IsWheelCollided(BYTE bWheel) = 0;
    virtual int  GetWheelFrictionState(BYTE bWheel) = 0;
```

File: Client/game_sa/CVehicleSA.h [hunk 4]
```diff
@@ -669,6 +669,7 @@ class CVehicleSA : public virtual CVehicle, public virtual CPhysicalSA
     void SetBikeWheelStatus(BYTE bWheel, BYTE bStatus);
 
     bool IsWheelCollided(BYTE eWheelPosition);
+    int GetWheelFrictionState(BYTE eWheelPosition);
```
Comment (qaisjp) [inline Client/game_sa/CVehicleSA.h]:
```suggestion
    int  GetWheelFrictionState(BYTE eWheelPosition);
```

**Review by qaisjp — CHANGES_REQUESTED** [review]
Excellent work. Please can you provide the example resource shown in the video? You can upload the zip into GitHub by dragging and dropping it into a comment (or into the edit box for your main issue).

**Comment by onikyannn** [comment]
@qaisjp Thank you.
Attached test resource to main issue.

**Review by qaisjp — APPROVED** [review]
Stability of mainline isn't that great right now but this should be available in a nightly tomorrow :)

Please update the wiki with docs!

**Comment by qaisjp** [comment]
I think we'll probably change this to return an enum string at some point

**Comment by Citizen01** [comment]
What about 2 wheels vehicles like motobikes and bikes ? I think their support should be added prior to release.

**Comment by onikyannn** [comment]
@Citizen01 tested with bike, don't work (always return 0).
I'll try to make some research about bikes.

**Comment by Citizen01** [comment]
yeah okay, actually I was expecting an error/warn in console, not 0 ?
https://github.com/multitheftauto/mtasa-blue/pull/1839/files#diff-9a175949acc865a4deea435d73c2082716ab68c6811ef1a657783f3d420dc00fR4120-R4121
```cpp
    if (CClientVehicleManager::GetVehicleType(pVehicle->GetModel()) != CLIENTVEHICLE_CAR)
        throw std::invalid_argument("Invalid vehicle type");
```
Thanks for the PR by the way, cool feature for the scripters to have 👍

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp [hunk 5]
```diff
@@ -4109,3 +4111,14 @@ bool CLuaVehicleDefs::SetVehicleModelWheelSize(const unsigned short usModel, con
...
     return true;
 }
+
+int CLuaVehicleDefs::GetVehicleWheelFrictionState(CClientVehicle* pVehicle, unsigned char wheel)
+{
+    if (wheel < 0 || wheel > 3)
+        throw std::invalid_argument("Invalid wheel number");
+
+    if (CClientVehicleManager::GetVehicleType(pVehicle->GetModel()) != CLIENTVEHICLE_CAR)
+        throw std::invalid_argument("Invalid vehicle type");
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp:4121]: I dont think the Lua API should be the one checking this. We should check it in `game_sa`, where we're `static_cast`ing it without actually making sure its an `automobile`

Comment (qaisjp) [reply Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp:4121]: What state should it return for non automobiles?

**Comment by qaisjp** [comment]
My guess, without reading the code, is that the wrong enum was used

**Comment by Citizen01** [comment]
> My guess, without reading the code, is that the wrong enum was used

Yeah maybe, but then who the hell named it like that. Looks like the perfect name if I wanted to target 4 wheels vehicles.

EDIT:
```cpp
enum eClientVehicleType
{
    CLIENTVEHICLE_NONE,
    CLIENTVEHICLE_CAR,
    CLIENTVEHICLE_BOAT,
    CLIENTVEHICLE_TRAIN,
    CLIENTVEHICLE_HELI,
    CLIENTVEHICLE_PLANE,
    CLIENTVEHICLE_BIKE,
    CLIENTVEHICLE_MONSTERTRUCK,
    CLIENTVEHICLE_QUADBIKE,
    CLIENTVEHICLE_BMX,
    CLIENTVEHICLE_TRAILER
};
```
Needs some investigation if it really returns 0 for a bike.
Anyway, seeing that enum, the following:
- CLIENTVEHICLE_MONSTERTRUCK,
- CLIENTVEHICLE_QUADBIKE,
- CLIENTVEHICLE_BMX

should be allowed too (not sure about the BMX though but as you can drift with it, I think it should be added to the supported list. Unless you run into problems with  that vehicle type, then it is fine to drop it)

**Comment by LosFaul** [comment]
just saying wiki entry is missing

**Comment by onikyannn** [comment]
@LosFaul thank you.
Have a lot of work now.
I plan to continue working on this task over the weekend.

**Comment by Pirulax** [comment]
iirc it should work with every class that inherits `CAutomobile`.

**Comment by Dutchman101** [comment]
@drop-club As expected for contributions like this, please document this function by adding it on the scripting wiki: [https://wiki.multitheftauto.com/wiki/GetVehicleWheelFrictionState](https://wiki.multitheftauto.com/wiki/GetVehicleWheelFrictionState) doesn't exist right now. Which means nobody would practically use it (then new functions go to waste)..

**Comment by LosFaul** [comment]
wiki entry still doesn't exists

**Comment by stoneage-mta** [comment]
I just added the wiki page, thanks to @xLuxy for helping.

**Comment by Yamsha75** [comment]
> Anyway, seeing that enum, the following:
> 
> * CLIENTVEHICLE_MONSTERTRUCK,
> * CLIENTVEHICLE_QUADBIKE,
> * CLIENTVEHICLE_BMX
> 
> should be allowed too (not sure about the BMX though but as you can drift with it, I think it should be added to the supported list. Unless you run into problems with that vehicle type, then it is fine to drop it)

Are these types confirmed to no work or were they just not tested and forgotten? The wiki does not mention anything about this function working only for "Automobile" vehicle types.

**Pull request merged**